File: scripts/generate_inventory.php
<?php
$root = dirname(__DIR__);
$dbRoot = $root . DIRECTORY_SEPARATOR . 'database';
$outPath = $dbRoot . DIRECTORY_SEPARATOR . 'INVENTAIRE_TABLES.md';

$sqlFiles = [];
$iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dbRoot));
foreach ($iterator as $file) {
    if ($file->isFile() && strtolower($file->getExtension()) === 'sql') {
        $sqlFiles[] = $file->getPathname();
    }
}
sort($sqlFiles);

function stripComments($text) {
    $lines = preg_split('/\R/', $text);
    $out = [];
    foreach ($lines as $line) {
        $out[] = preg_replace('/--.*$/', '', $line);
    }
    return implode("\n", $out);
}

function normalizeType($defn) {
    $s = trim($defn);
    $s = preg_replace('/\bNOT NULL\b/i', '', $s);
    $s = preg_replace('/\bNULL\b/i', '', $s);
    $s = preg_replace('/\bPRIMARY KEY\b/i', '', $s);
    $s = preg_replace('/\bUNIQUE\b/i', '', $s);
    $s = preg_replace('/\bAUTO_INCREMENT\b/i', '', $s);
    $s = preg_replace('/\bDEFAULT\s+[^,\s]+/i', '', $s);
    $s = preg_replace('/\bON UPDATE CURRENT_TIMESTAMP\b/i', '', $s);
    $s = preg_replace('/\bCHARACTER SET[^,]+/i', '', $s);
    $s = preg_replace('/\bCOLLATE[^,]+/i', '', $s);
    $s = preg_replace('/\s+/', ' ', $s);
    return trim(rtrim($s, ','));
}

function normalizeDefault($defn) {
    if (preg_match('/DEFAULT\s+([^\s,]+)/i', $defn, $m)) {
        $value = strtoupper($m[1]);
        if ($value === 'CURRENT_TIMESTAMP') return 'CURRENT_TIMESTAMP';
        if ($value === 'NULL') return 'NULL';
        return $m[1];
    }
    return (stripos($defn, 'NULL') !== false) ? 'NULL' : '—';
}

function normalizeNullable($defn) {
    return stripos($defn, 'NOT NULL') !== false ? 'NO' : 'YES';
}

function normalizeKey($defn, $colName, $fkCols) {
    $upper = strtoupper($defn);
    if (strpos($upper, 'PRIMARY KEY') !== false) return 'PRI';
    if (strpos($upper, 'UNIQUE') !== false) return 'UNI';
    if (in_array($colName, $fkCols, true) || substr($colName, -3) === '_id') return 'MUL';
    return '—';
}

function normalizeExtra($defn) {
    $upper = strtoupper($defn);
    $extra = strpos($upper, 'AUTO_INCREMENT') !== false ? 'auto_increment' : '—';
    if (strpos($upper, 'ON UPDATE CURRENT_TIMESTAMP') !== false) {
        return $extra === '—' ? 'DEFAULT_GENERATED on update CURRENT_TIMESTAMP' : $extra . ' on update CURRENT_TIMESTAMP';
    }
    if (strpos($upper, 'DEFAULT') !== false && strpos($upper, 'CURRENT_TIMESTAMP') !== false && $extra === '—') {
        return 'DEFAULT_GENERATED';
    }
    return $extra;
}

function parseColumns($body) {
    $lines = preg_split('/\R/', $body);
    $columns = [];
    $fkCols = [];
    foreach ($lines as $rawLine) {
        $line = trim($rawLine);
        if ($line === '') continue;
        if (preg_match('/^(CONSTRAINT|PRIMARY|UNIQUE|KEY|INDEX|CHECK)/i', $line)) {
            if (preg_match('/FOREIGN\s+KEY\s*\(([^)]+)\)/i', $line, $m)) {
                foreach (preg_split('/,/', $m[1]) as $col) {
                    $fkCols[] = trim(str_replace('`', '', $col));
                }
            }
            continue;
        }
        if (preg_match('/^FOREIGN/i', $line)) {
            if (preg_match('/FOREIGN\s+KEY\s*\(([^)]+)\)/i', $line, $m)) {
                foreach (preg_split('/,/', $m[1]) as $col) {
                    $fkCols[] = trim(str_replace('`', '', $col));
                }
            }
            continue;
        }
        if ($line === ')' || $line === ');' || $line === '},') continue;
        if (preg_match('/^`?([a-zA-Z0-9_]+)`?\s+(.+)$/', $line, $m)) {
            $colName = $m[1];
            $colDef = trim(rtrim($m[2], ','));
            $columns[] = [
                'name' => $colName,
                'definition' => $colDef,
                'type' => normalizeType($colDef),
                'null' => normalizeNullable($colDef),
                'key' => normalizeKey($colDef, $colName, $fkCols),
                'default' => normalizeDefault($colDef),
                'extra' => normalizeExtra($colDef),
            ];
        }
    }
    return $columns;
}

$schemaTables = [];
foreach ($sqlFiles as $filePath) {
    $text = file_get_contents($filePath);
    $clean = stripComments($text);

    preg_match_all('/CREATE\s+TABLE(?:\s+IF\s+NOT\s+EXISTS)?\s+`?([a-zA-Z0-9_]+)`?\s*\((.*?)\);/is', $clean, $matches, PREG_SET_ORDER);
    foreach ($matches as $match) {
        $tableName = strtolower($match[1]);
        if ($tableName === 'sqlite_sequence') continue;
        $body = $match[2];
        $schemaTables[$tableName] = [
            'name' => $tableName,
            'columns' => parseColumns($body),
        ];
    }

    preg_match_all('/ALTER\s+TABLE\s+`?([a-zA-Z0-9_]+)`?\s+(.+?);/is', $clean, $alterMatches, PREG_SET_ORDER);
    foreach ($alterMatches as $alter) {
        $tableName = strtolower($alter[1]);
        if (!isset($schemaTables[$tableName])) {
            $schemaTables[$tableName] = ['name' => $tableName, 'columns' => []];
        }
        $body = $alter[2];

        if (stripos($body, 'ADD COLUMN') !== false) {
            preg_match_all('/ADD\s+COLUMN(?:\s+IF\s+NOT\s+EXISTS)?\s+`?([a-zA-Z0-9_]+)`?\s+(.+?)(?=(?:,\s+ADD\s+COLUMN)|$)/is', $body, $addMatches, PREG_SET_ORDER);
            foreach ($addMatches as $add) {
                $colName = $add[1];
                $colDef = trim(rtrim($add[2], ','));
                $colDef = preg_replace('/\b(AFTER|FIRST)\s+`?[a-zA-Z0-9_]+`?$/i', '', $colDef);
                $colDef = trim($colDef);
                $columns = &$schemaTables[$tableName]['columns'];
                $existing = null;
                foreach ($columns as &$col) {
                    if ($col['name'] === $colName) {
                        $existing = &$col;
                        break;
                    }
                }
                if ($existing) {
                    $existing['definition'] = $colDef;
                    $existing['type'] = normalizeType($colDef);
                    $existing['null'] = normalizeNullable($colDef);
                    $existing['key'] = normalizeKey($colDef, $colName, []);
                    $existing['default'] = normalizeDefault($colDef);
                    $existing['extra'] = normalizeExtra($colDef);
                } else {
                    $columns[] = [
                        'name' => $colName,
                        'definition' => $colDef,
                        'type' => normalizeType($colDef),
                        'null' => normalizeNullable($colDef),
                        'key' => normalizeKey($colDef, $colName, []),
                        'default' => normalizeDefault($colDef),
                        'extra' => normalizeExtra($colDef),
                    ];
                }
                unset($col);
                unset($existing);
            }
        }

        if (stripos($body, 'MODIFY COLUMN') !== false) {
            preg_match_all('/MODIFY\s+COLUMN\s+`?([a-zA-Z0-9_]+)`?\s+(.+?)(?=(?:,\s+ADD\s+COLUMN)|$)/is', $body, $modMatches, PREG_SET_ORDER);
            foreach ($modMatches as $mod) {
                $colName = $mod[1];
                $colDef = trim(rtrim($mod[2], ','));
                $colDef = preg_replace('/\b(AFTER|FIRST)\s+`?[a-zA-Z0-9_]+`?$/i', '', $colDef);
                $colDef = trim($colDef);
                $columns = &$schemaTables[$tableName]['columns'];
                foreach ($columns as &$col) {
                    if ($col['name'] === $colName) {
                        $col['definition'] = $colDef;
                        $col['type'] = normalizeType($colDef);
                        $col['null'] = normalizeNullable($colDef);
                        $col['key'] = normalizeKey($colDef, $colName, []);
                        $col['default'] = normalizeDefault($colDef);
                        $col['extra'] = normalizeExtra($colDef);
                        break;
                    }
                }
                unset($col);
            }
        }
    }
}

$lines = [];
$lines[] = '# Inventaire de la base de données medecin';
$lines[] = '';
$lines[] = 'Généré automatiquement à partir des scripts SQL et migrations présents dans le dépôt.';
$lines[] = '';
$lines[] = 'Nombre de tables répertoriées : ' . count($schemaTables);
$lines[] = '';

ksort($schemaTables);
foreach ($schemaTables as $tableName => $item) {
    $columns = $item['columns'];
    $lines[] = '## `' . $tableName . '` (' . count($columns) . ' champs)';
    $lines[] = '';
    $lines[] = '| Champ | Type | Null | Clé | Défaut | Extra |';
    $lines[] = '|-------|------|------|-----|--------|-------|';
    foreach ($columns as $col) {
        $lines[] = '| `' . $col['name'] . '` | ' . $col['type'] . ' | ' . $col['null'] . ' | ' . $col['key'] . ' | ' . $col['default'] . ' | ' . $col['extra'] . ' |';
    }
    $lines[] = '';
}

file_put_contents($outPath, implode("\n", $lines));
echo "Wrote $outPath with " . count($schemaTables) . " tables\n";
?>

File: scripts/generate_inventory.py
import re
import pathlib
from collections import OrderedDict

ROOT = pathlib.Path(r'c:\wamp64\www\medecin')
DB_ROOT = ROOT / 'database'
OUT_PATH = DB_ROOT / 'INVENTAIRE_TABLES.md'

CREATE_TABLE_RE = re.compile(r'CREATE\s+TABLE(?:\s+IF\s+NOT\s+EXISTS)?\s+`?([a-zA-Z0-9_]+)`?\s*\((.*?)\);', re.I | re.S)
ALTER_TABLE_RE = re.compile(r'ALTER\s+TABLE\s+`?([a-zA-Z0-9_]+)`?\s+(.+?);', re.I | re.S)
ADD_COLUMN_RE = re.compile(r'ADD\s+COLUMN(?:\s+IF\s+NOT\s+EXISTS)?\s+`?([a-zA-Z0-9_]+)`?\s+(.+?)(?=(?:,\s+ADD\s+COLUMN)|$)', re.I | re.S)
MODIFY_COLUMN_RE = re.compile(r'MODIFY\s+COLUMN\s+`?([a-zA-Z0-9_]+)`?\s+(.+?)(?=(?:,\s+ADD\s+COLUMN)|$)', re.I | re.S)


def strip_comments(text: str) -> str:
    lines = []
    for raw in text.splitlines():
        line = re.sub(r'--.*$', '', raw)
        lines.append(line)
    return '\n'.join(lines)


def normalize_type(defn: str) -> str:
    s = defn.strip()
    s = re.sub(r'\bNOT NULL\b', '', s, flags=re.I)
    s = re.sub(r'\bNULL\b', '', s, flags=re.I)
    s = re.sub(r'\bPRIMARY KEY\b', '', s, flags=re.I)
    s = re.sub(r'\bUNIQUE\b', '', s, flags=re.I)
    s = re.sub(r'\bAUTO_INCREMENT\b', '', s, flags=re.I)
    s = re.sub(r'\bDEFAULT\s+[^,\s]+', '', s, flags=re.I)
    s = re.sub(r'\bON UPDATE CURRENT_TIMESTAMP\b', '', s, flags=re.I)
    s = re.sub(r'\bCHARACTER SET[^,]+', '', s, flags=re.I)
    s = re.sub(r'\bCOLLATE[^,]+', '', s, flags=re.I)
    s = re.sub(r'\s+', ' ', s).strip().rstrip(',')
    return s


def normalize_default(defn: str):
    m = re.search(r'DEFAULT\s+([^\s,]+)', defn, flags=re.I)
    if m:
        value = m.group(1)
        if value.upper() == 'CURRENT_TIMESTAMP':
            return 'CURRENT_TIMESTAMP'
        if value.upper() == 'NULL':
            return 'NULL'
        return value
    return 'NULL' if 'NULL' in defn.upper() else '—'


def normalize_nullable(defn: str) -> str:
    return 'NO' if 'NOT NULL' in defn.upper() else 'YES'


def normalize_key(defn: str, col_name: str, fk_cols: set) -> str:
    upper = defn.upper()
    if 'PRIMARY KEY' in upper:
        return 'PRI'
    if 'UNIQUE' in upper:
        return 'UNI'
    if col_name in fk_cols or col_name.endswith('_id'):
        return 'MUL'
    return '—'


def normalize_extra(defn: str) -> str:
    upper = defn.upper()
    extra = 'auto_increment' if 'AUTO_INCREMENT' in upper else '—'
    if 'ON UPDATE CURRENT_TIMESTAMP' in upper:
        if extra == '—':
            return 'DEFAULT_GENERATED on update CURRENT_TIMESTAMP'
        return extra + ' on update CURRENT_TIMESTAMP'
    if 'DEFAULT' in upper and 'CURRENT_TIMESTAMP' in upper and extra == '—':
        return 'DEFAULT_GENERATED'
    return extra


def parse_columns_from_body(body: str):
    lines = [ln.strip() for ln in body.splitlines() if ln.strip()]
    columns = []
    fk_cols = set()
    for line in lines:
        if not line:
            continue
        if line.startswith('CONSTRAINT') or line.startswith('PRIMARY') or line.startswith('UNIQUE') or line.startswith('KEY') or line.startswith('INDEX') or line.startswith('CHECK'):
            fk_match = re.search(r'FOREIGN\s+KEY\s*\(([^)]+)\)', line, re.I)
            if fk_match:
                cols = [c.strip().strip('`') for c in fk_match.group(1).split(',')]
                fk_cols.update(cols)
            continue
        if line.startswith('FOREIGN'):
            fk_match = re.search(r'FOREIGN\s+KEY\s*\(([^)]+)\)', line, re.I)
            if fk_match:
                cols = [c.strip().strip('`') for c in fk_match.group(1).split(',')]
                fk_cols.update(cols)
            continue
        if line in {')', ');', '},'}:
            continue
        col_match = re.match(r'`?([a-zA-Z0-9_]+)`?\s+(.+)$', line)
        if not col_match:
            continue
        col_name = col_match.group(1)
        col_def = col_match.group(2).strip().rstrip(',')
        columns.append({
            'name': col_name,
            'definition': col_def,
            'type': normalize_type(col_def),
            'null': normalize_nullable(col_def),
            'key': normalize_key(col_def, col_name, fk_cols),
            'default': normalize_default(col_def),
            'extra': normalize_extra(col_def),
        })
    return columns


sql_files = sorted([p for p in DB_ROOT.rglob('*.sql') if p.is_file()])

schema_tables = OrderedDict()
for path in sql_files:
    text = path.read_text(encoding='utf-8-sig')
    clean = strip_comments(text)
    for match in CREATE_TABLE_RE.finditer(clean):
        table_name = match.group(1).lower()
        if table_name in {'sqlite_sequence'}:
            continue
        body = match.group(2)
        columns = parse_columns_from_body(body)
        schema_tables[table_name] = {'name': table_name, 'columns': columns}

    for match in ALTER_TABLE_RE.finditer(clean):
        table_name = match.group(1).lower()
        body = match.group(2)
        if table_name not in schema_tables:
            schema_tables[table_name] = {'name': table_name, 'columns': []}
        if 'ADD COLUMN' in body.upper():
            for add in ADD_COLUMN_RE.finditer(body):
                col_name = add.group(1)
                col_def = add.group(2).strip().rstrip(',')
                col_def = re.sub(r'\b(AFTER|FIRST)\s+`?[a-zA-Z0-9_]+`?$', '', col_def, flags=re.I).strip()
                columns = schema_tables[table_name]['columns']
                existing = next((c for c in columns if c['name'] == col_name), None)
                if existing:
                    existing.update({
                        'definition': col_def,
                        'type': normalize_type(col_def),
                        'null': normalize_nullable(col_def),
                        'key': normalize_key(col_def, col_name, set()),
                        'default': normalize_default(col_def),
                        'extra': normalize_extra(col_def),
                    })
                else:
                    columns.append({
                        'name': col_name,
                        'definition': col_def,
                        'type': normalize_type(col_def),
                        'null': normalize_nullable(col_def),
                        'key': normalize_key(col_def, col_name, set()),
                        'default': normalize_default(col_def),
                        'extra': normalize_extra(col_def),
                    })
        if 'MODIFY COLUMN' in body.upper():
            for mod in MODIFY_COLUMN_RE.finditer(body):
                col_name = mod.group(1)
                col_def = mod.group(2).strip().rstrip(',')
                col_def = re.sub(r'\b(AFTER|FIRST)\s+`?[a-zA-Z0-9_]+`?$', '', col_def, flags=re.I).strip()
                columns = schema_tables[table_name]['columns']
                existing = next((c for c in columns if c['name'] == col_name), None)
                if existing:
                    existing.update({
                        'definition': col_def,
                        'type': normalize_type(col_def),
                        'null': normalize_nullable(col_def),
                        'key': normalize_key(col_def, col_name, set()),
                        'default': normalize_default(col_def),
                        'extra': normalize_extra(col_def),
                    })

lines = []
lines.append('# Inventaire de la base de données medecin')
lines.append('')
lines.append(f'Généré automatiquement à partir des scripts SQL et migrations présents dans le dépôt le {ROOT.name}.')
lines.append('')
lines.append(f'Nombre de tables répertoriées : {len(schema_tables)}')
lines.append('')

for table_name, item in sorted(schema_tables.items(), key=lambda kv: kv[0]):
    columns = item['columns']
    lines.append(f'## `{table_name}` ({len(columns)} champs)')
    lines.append('')
    lines.append('| Champ | Type | Null | Clé | Défaut | Extra |')
    lines.append('|-------|------|------|-----|--------|-------|')
    for col in columns:
        lines.append(f"| `{col['name']}` | {col['type']} | {col['null']} | {col['key']} | {col['default']} | {col['extra']} |")
    lines.append('')

OUT_PATH.write_text('\n'.join(lines), encoding='utf-8')
print(f'Wrote {OUT_PATH} with {len(schema_tables)} tables')
